src/components/ListOfGoods/ListOfGoods.tsx

```diff
@@ -0,0 +1,110 @@
+import * as React from 'react';
+
+type Props = {
+  list: string[],
+};
+
+type State = {
+  isReversed: boolean;
+  isSortedBy: 'name' | 'length' | 'none';
+  reset: boolean;
+};
+
+class ListOfGoods extends React.Component<Props, State> {
+  state: State = {
+    isReversed: false,
+    isSortedBy: 'none',
+    reset: false,
+  };
+
+  reverseList = () => {
+    this.setState(state => ({
+      isReversed: !state.isReversed,
+    }));
+  };
+
+  sortListByName = () => {
+    this.setState((currentState) => ({
+      isSortedBy: currentState.isSortedBy === 'name'
+        ? 'none'
+        : 'name',
+    }));
+  };
+
+  resetList = () => {
+    this.setState(state => ({
+      reset: !state.reset,
+    }));
+  };
+
+  sortByLengthList = () => {
+    this.setState((currentState) => ({
+      isSortedBy: currentState.isSortedBy === 'length'
+        ? 'none'
+        : 'length',
+    }));
+  };
+
+  sorting = (newList: string[]) => {
+    switch (this.state.isSortedBy) {
+      case 'name':
+        newList.sort((firsGood, secondGood) => firsGood.localeCompare(secondGood));
+        break;
+
+      case 'length':
+        newList.sort((firsGood, secondGood) => firsGood.length - secondGood.length);
+        break;
+
+      default:
+        break;
+    }
+  };
+
+  reversing = (newList: string[]) => {
+    newList.reverse();
+  };
+
+  resetting = () => {
+    this.state.reset = false;
+    this.state.isSortedBy = 'none';
+    this.state.isReversed = false;
+  };
+
+  render() {
+    const { list } = this.props;
+    let newList = [...list];
+
+    if (this.state.isSortedBy !== 'none') {
+      this.sorting(newList);
+    }
+
+    if (this.state.isReversed) {
+      this.reversing(newList);
+    }
+
+    if (this.state.reset) {
+      newList = [...list];
+      this.resetting();
+    }
```

Comment:
all sortings and reseting etc should be done on methods, not in render. Also when you click 'sort' a few times - goods shouldn't reset. Reset occurs only by pressing on 'reset' button